### ISSUE TITLE
pywinauto: how to enter data on a ListView 

### ISSUE THREAD

**Issue by jagadeesh1983** [pywinauto/pywinauto#410]
Hi,

I have an application which has ListView Control and I need to enter data into it.
But as per PyWinAuto documentation. There is no commands or API to enter data for ListView Control.
http://pywinauto.readthedocs.io/en/latest/code/pywinauto.controls.common_controls.html?highlight=SysListView
Please find below the screenshot of the application which shows List View Control.
I have manually entered data into 1st row  but i need to enter that data through pywinauto code.
![screenshot](https://user-images.githubusercontent.com/29576279/30424925-0855246e-9965-11e7-8328-fccc3244c95b.jpg)

Please let me know how to enter data into ListView Control items.

Thanks,
Jagadeesh

**Comment by vasily-v-ryabov** [comment]
Hi @jagadeesh1983, you can use something like that:
```python
item = list_view.get_item(index_or_str)
item.click_input(double=True)
list_view.type_keys('something{ENTER}', set_foreground=False)
```
The second param is to prevent changing focus.

**Comment by jagadeesh1983** [comment]
Hi Vasily,
I am getting following error on trying your above mentioned code.

{'text': u'1', 'state': None, 'indent': 0, 'image': 0}
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Work\pywinauto\Script2.py", line 1, in <module>
    from pywinauto import application
AttributeError: 'dict' object has no attribute 'click_input'

My Code snippet:
item = app.window_(title_re = 'Analytic Type Transaction Property').ListView3.GetItem(0)
print item
item.click_input(double=True)
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.type_keys('Hi{ENTER}', set_foreground=False)

I have tried using all combinations like click_input(double=True), Click_Input(double=True),ClickInput(double=True) and clickInput(double=True) but its of no use. All giving the same error.
![screenshot1](https://user-images.githubusercontent.com/29576279/30445343-52b924b8-99a3-11e7-85aa-983a94f68747.jpg)

Also if I just comment out the line item.click_input(double=True) and execute the rest of the above code the I get following error:
>>> {'text': u'1', 'state': None, 'indent': 0, 'image': 0}
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Work\pywinauto\Script2.py", line 1, in <module>
    from pywinauto import application
  File "C:\Python27\lib\site-packages\pywinauto\application.py", line 234, in __getattr__
    return getattr(ctrls[-1], attr)
AttributeError: 'ListViewWrapper' object has no attribute 'type_keys'

Basically I am getting AttributeError for both click_input() and type_keys().

Please help.

**Comment by vasily-v-ryabov** [comment]
Probably you use old pywinauto (I remember you mentioned 0.4.0 or 0.4.2). Sorry, it very old version and works on 32-bit Python only (BTW, if you use 64-bit Python, 0.5.0 is a minimum, but 0.5.4 is much better). It's really easier to fix compatibility problems in 0.6.x releases (if you have them) than supporting 0.4.x.

What I can say this line won't work for 0.4.2 because param `set_foreground` was introduced in 0.5.0:
```python
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.type_keys('Hi{ENTER}', set_foreground=False)
```

So which exact pywinauto version do you use for this task? Is it 32-bit or 64-bit Python and app?

**Comment by jagadeesh1983** [comment]
We are using Python 2.7 (64 bit) with pyWinAuto 0.4 (64 bit) for our automation project.
Here the issue is not with param set_foreground. I am getting AttributeError for both click_input() and type_keys() even if I don't use param set_foreground in type_keys() method. 

Even if I try execute the following code also getting Atrribute error for click_input():
Code snippet:
item = app.window_(title_re = 'Analytic Type Transaction Property').ListView3.GetItem(0)
print item
item.click_input(double=True)

Error:
{'text': u'1', 'state': None, 'indent': 0, 'image': 0}
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Work\pywinauto\Script2.py", line 1, in <module>
    from pywinauto import application
AttributeError: 'dict' object has no attribute 'ClickInput'

**Comment by vasily-v-ryabov** [comment]
Sorry I don't understand what 0.4 means. Can you provide all 3 numbers including minor one? Is it 0.5.4?

If yes, then just use `ClickInput` instead of `click_input` and `TypeKeys` instead of `type_keys`.

**Comment by vasily-v-ryabov** [comment]
And try `Item(0)` instead of `GetItem(0)`. Because `GetItem(0)` returns a dictionary of item properties, not item object itself.

**Comment by jagadeesh1983** [comment]
We are using pyWinAuto 0.4.1 (64 bit).
I have tried using Item(0) instead of GetItem(0) as per your suggestion but getting this Attribute error:

Code Snippet:
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.Item(0)

OutPut:
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Work\pywinauto\Script2.py", line 1, in <module>
    from pywinauto import application
  File "C:\Python27\lib\site-packages\pywinauto\application.py", line 234, in __getattr__
    return getattr(ctrls[-1], attr)
AttributeError: 'ListViewWrapper' object has no attribute 'Item'

**Comment by jagadeesh1983** [comment]
Once the above issue is fixed for Item() then once I can able to try ClickInput() and TypeKeys to enter data into row in ListView item.

**Comment by vasily-v-ryabov** [comment]
OK, if you mean [this 64-bit version of pywinauto](https://github.com/mpaf/pywinauto-64bit) with removed asserts, it may partially work but we can't guarantee it works for 64-bit apps. Because many internal structure fields have incorrect size that may cause your app to crash or unpredictable behavior. Official release 0.4.0 or 0.4.2 (made by Mark in 2010) will even not run on 64-bit Python. So minimal version for 64-bit apps we support is pywinauto 0.5.4. Consider upgrading to this or newest version. Or do it for your own risk with your own modifications (it's open source after all).

Sorry, I've rechecked now: `GetItem(0)` is correct method. It should return item object in 0.5.4.
But in 0.4.1 this item object class even doesn't exist. Because we implemented it in 0.5.x releases specially for ListView (to make it similar to TreeView items that exist in 0.4.x).

**Comment by jagadeesh1983** [comment]
Thanks much for the detailed explanation.
Sure, as per your recommendation I will upgrade to newest version of pyWinAuto and will try to use your above code to enter data into ListView.

**Comment by jagadeesh1983** [comment]
I have upgrade to new pywinauto version 0.6.3 and tried your above mentioned code to enter data. 
But this time there was no error or exception but it is still not entering data into cell for List View.
My Code Snippet:
app.window_(title_re = 'Analytic Type Transaction Property').SetFocus()
item = app.window_(title_re = 'Analytic Type Transaction Property').ListView3.GetItem(0)
item.click_input(double=True)
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.type_keys('Hi{ENTER}', set_foreground=False)

Also I saw pywinauto 0.6.3 documentation and I found there is a method called cell(row, column) for ListView control which will return a cell in the ListView control.
I thought this cell()method would be idle for me to get the access of the cell to enter data but it was giving following error when I execute this code.

Error:
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Users\jkanthar\Documents\Script2.py", line 1, in <module>
    from pywinauto import application
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\application.py", line 173, in __call__
    format(self.criteria[-1]['best_match']))
AttributeError: WindowSpecification class has no 'cell' method

Please let me know how to resolve this issue to access cell() method

**Comment by vasily-v-ryabov** [comment]
Hi @jagadeesh1983 can you add some pause after double click on the cell? Like this:
```python
app.window_(title_re = 'Analytic Type Transaction Property').SetFocus()
item = app.window_(title_re = 'Analytic Type Transaction Property').ListView3.GetItem(0)
item.click_input(double=True)
time.sleep(1) # wait 1 sec.
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.type_keys('Hi{ENTER}', set_foreground=False)
```
Getting to edit mode takes some time probably.

If you want to use method `.cell()` it's inside new backend (works through UI Automation API since 0.6.0; your current code uses default Win32 API backend inherited from 0.5.4). To switch backend use `Application(backend="uia")`. And please read the [Getting Started Guide](https://pywinauto.readthedocs.io/en/latest/getting_started.html) carefully since you mention the docs. Spy/Inspect tools should help you a lot to decide what is more suitable for you.

**Comment by jagadeesh1983** [comment]
I have tried including wait statement ie time.sleep(3) after double click operation but still the same issue.
It does not click the cell instead it clicks the 1st Row due to this its not able to enter the data into the particular cell.

Regarding usage of Application(backend="uia"), I have tried the following code to use new backend.
app = application.Application(backend="uia").connect(title='Application Designer - Untitled') 

But I was getting following error:
Exception WindowsError: 'exception: access violation writing 0x0000000000040000' in <bound method tagVARIANT.__del__ of VARIANT(vt=0x0, None)> ignored
Traceback (most recent call last):
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\framework\scriptutils.py", line 323, in RunScript
    debugger.run(codeObject, __main__.__dict__, start_stepping=0)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\__init__.py", line 60, in run
    _GetCurrentDebugger().run(cmd, globals,locals, start_stepping)
  File "C:\Python27\Lib\site-packages\pythonwin\pywin\debugger\debugger.py", line 655, in run
    exec cmd in globals, locals
  File "C:\Users\jkanthar\Documents\Script2.py", line 1, in <module>
    from pywinauto import application
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\application.py", line 944, in connect
    self.process = findwindows.find_element(**kwargs).process_id
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\findwindows.py", line 84, in find_element
    elements = find_elements(**kwargs)
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\findwindows.py", line 197, in find_elements
    cache_enable=True)
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\uia_element_info.py", line 271, in children
    cond = IUIA().build_condition(**kwargs)
  File "C:\Python27\lib\site-packages\pywinauto-0.6.3-py2.7.egg\pywinauto\uia_defines.py", line 113, in build_condition
    conditions.append(self.iuia.CreatePropertyCondition(self.UIA_dll.UIA_NamePropertyId, title))
ArgumentError: argument 2: <type 'exceptions.WindowsError'>: exception: access violation writing 0x0000000000040000

I am not sure if my application supports (backend="uia") options.

**Comment by jagadeesh1983** [comment]
item.click_input(double=True) command clicks on the row inside of cell in a particular.
is there any command to click on particular cell in a row.

**Comment by vasily-v-ryabov** [comment]
Access violation issue is a known problem in `comtypes` package I submitted a while ago. It seems this is not resolved yet.

See https://github.com/enthought/comtypes/issues/98

The issue with clicking on the first cell in a row could be workarounded, but need a bit more time to find the solution. Just try to list all items using `.items()` method to check which index is appropriate for target column.

Usually enumeration of cells goes this way:
| 0 | 5 | 10 |
| 1 | 6 | 11 |
| 2 | 7 | 12 |
| 3 | 8 | 13 |
| 4 | 9 | 14 |

**Comment by vasily-v-ryabov** [comment]
comtypes problems happens on my machine if I'm completely re-installing Python (even the same location) and comtypes was used successfully in previous installation. No idea what to clear in the system to make it working again.

**Comment by jagadeesh1983** [comment]
I tried to list all the items using .items() method as per your suggestion but it is not giving the index numbers of the cell.
My Code snippet:
item =  app.window_(title_re = 'Analytic Type Transaction Property').ListView3.Items()
print item

Output:

[{'text': u'1', 'state': None, 'indent': 0, 'image': 0}, {'text': u'', 'state': None, 'indent': 0, 'image': 0}, {'text': u'', 'state': None, 'indent': 0, 'image': 0}, {'text': u'', 'state': None, 'indent': 0, 'image': 0}, {'text': u'', 'state': None, 'indent': 0, 'image': 0}, {'text': u'', 'state': None, 'indent': 0, 'image': 0}]

Please let me know if you are able to find any workaround for clicking on the first cell in a row and entering data.

**Comment by jagadeesh1983** [comment]
I have got the workaround to enter the data into the cell for ListView item.
Here is my code which will enter data into cell:
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.ClickInput(button='Left',coords=(115, 32))
app.window_(title_re ='Analytic Type Transaction Property').TypeKeys('TEST')

In this case, I have make use of co-ordinates to click on the 2nd cell in 1st Row.

We can even make some calculation to find out the exact co-ordinates ie x and y axis of particular cell in a row using below 3 codes:

app.window_(title_re = 'Analytic Type Transaction Property').ListView3.ColumnCount()
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.ColumnWidths()
app.window_(title_re = 'Analytic Type Transaction Property').ListView3.GetItemRect(0) 

Thanks for your help and support.
Please close this ticket.

**Comment by vasily-v-ryabov** [comment]
Hi @jagadeesh1983 this is great! We will think how to simplify this. So I'd prefer to keep it open. Thanks for your patience!

**Comment by vasily-v-ryabov** [comment]
This is implemented in 0.6.4.

Examples:
```python
    auto_detected_ctrl = list_view.get_item(0).inplace_control()

    combo = list_view.get_item(1,1).inplace_control("ComboBox")
    combo.select("Item name")

    edit = list_view.get_item(3,4).inplace_control("Edit")
    edit.type_keys("some text{ENTER}", set_foreground=False)

    dt_picker = list_view.get_item(2,0).inplace_control("DateTimePicker")
```
